### ISSUE TITLE
websockets connections fail

### ISSUE THREAD

**Issue by MarneeDear** [SAFE-Stack/SAFE-template#206]
I followed the [Quickstart Guide](https://safe-stack.github.io/docs/quickstart/). The app loads and appears to work (the counter works), but the websockets connections are failing (as seen through Chrome dev tools). 

I think it is the port. The server runs on port `8085`, but the websockets are trying to connect on port `8000`, and the server doesn't have a `socketcluster` route.

```
WebSocket connection to 'ws://localhost:8000/socketcluster/' failed: Connection closed before receiving a handshake response
```

## Other details
Windows 10
dotnet 2.2.101
fake 5.11.1
yarn 1.12.1
node 10.13

**Comment by Zaid-Ajaj** [comment]
Hello @MarneeDear I think you need to setup the proxy of the webpack-dev-server to re-route websocket requests to the actual backend, here is an example, see second entry of `proxy`:
```js
devServer: {
    proxy: {
      '/api/*': {
        target: 'http://localhost:8085',
        changeOrigin: true
      },
      '/socket': {
        target: 'http://localhost:8085',
        ws: true
     }
    },
    contentBase: "./public",
    hot: true,
    inline: true
  }
```

**Comment by MarneeDear** [comment]
@Zaid-Ajaj Thank you but that seems to already be the case. webpack is configured to get the server port from `process.env.SERVER_PROXY_PORT` or `8085`.  

Can you tell me where or how `process.env.SERVER_PROXY_PORT` is configured?

 This is what my webpack config looks like. This is what came with the template.

```
    // Configuration for the development server
    devServer: {
        // redirect requests that start with /api/* to the server on port 8085
        proxy: {
            '/api/*': {
                target: 'http://localhost:' + (process.env.SERVER_PROXY_PORT || "8085"),
                changeOrigin: true
            }
        },
```

**Comment by Zaid-Ajaj** [comment]
AFAIK, the `/api/*` entry of the proxy re-routes only requests with `http://` scheme (websockets use `ws://` scheme).

Can you please try changing your config into the following: 
```
    // Configuration for the development server
    devServer: {
        // redirect requests that start with /api/* to the server on port 8085
        proxy: {
            '/api/*': {
                target: 'http://localhost:' + (process.env.SERVER_PROXY_PORT || "8085"),
                changeOrigin: true
            }, 
            '/socketcluster': {
                 target: 'http://localhost:' + (process.env.SERVER_PROXY_PORT || "8085"),
                 ws: true
             }
        },
``` 
notice the property `ws: true` that tell webpack dev server that this is for web sockets

**Comment by Nhowka** [comment]
Probably you need `"ws" : true` to proxy the websockets.

Em sex, 28 de dez de 2018 15:16, Marnee Dearman (KG7SIO) <
notifications@github.com escreveu:

**Comment by MarneeDear** [comment]
Thank you @Zaid-Ajaj .

I tried adding the socketcluster config to my webpack config but I am `still getting the same websocket error`. I stopped the server and ran the build a few times just to make sure.

My webpack config `devServer` config looks like this now:

```
devServer: {
        // redirect requests that start with /api/* to the server on port 8085
        proxy: {
            '/api/*': {
                target: 'http://localhost:' + (process.env.SERVER_PROXY_PORT || "8085"),
                changeOrigin: true
            },
            '/socketcluster': {
                target: 'http://localhost:' + (process.env.SERVER_PROXY_PORT || "8085"),
                ws: true
            }
        },        
        // turn on hot module reloading
        hot: true,
        // more automatic reloading
        inline: true,
        // default page
        historyApiFallback: { index: resolve("./index.html") },
        // where to server static files from
        contentBase: resolve("./public")
    },
```

**Comment by isaacabraham** [comment]
I think that this might be a false positive - I seem to recall this can be fixed by installing the Redux Chrome extension (or perhaps the React one). Can you try installing all the prereqs here:

https://safe-stack.github.io/docs/feature-debugging/#prerequisites

And seeing if that fixes it?

**Comment by isaacabraham** [comment]
@Zaid-Ajaj if there's an issue with the webpack config that will make this go away, could you submit a PR to the template? :-)

**Comment by MarneeDear** [comment]
@isaacabraham 

Installing `Redux Dev Tools` and `React Developer Tools` seemed to have worked. I am not seeing a WebSocket error anymore.

I tried removing the `socketcluster`  from the `devServer` proxy and the websocket connection error didn't come back, so I dont think it is needed, but Im not sure.

Thanks everyone for all of your quick help!

@Zaid-Ajaj Thank your for your explanation. That helped me a lot to understand how this all works.

**Comment by isaacabraham** [comment]
We should probably add this to the documentation somewhere.

**Comment by MarneeDear** [comment]
I was thinking I could do that. Would a note in the Troubleshooting section of the Quickstart Guide make sense?

**Comment by MarneeDear** [comment]
@isaacabraham Do you think the error can be safely ignore?

**Comment by Zaid-Ajaj** [comment]
@MarneeDear glad you got it working!

Only it seems a bit weird, I just looked into the [code](https://github.com/webpack/webpack-dev-server/blob/master/lib/Server.js#L356) of webpack-dev-server, doesn't seem to proxy websockets unless there is `ws: true` property. I will try test things later and will open an issue if I find anything.

**Comment by isaacabraham** [comment]
@marneedear Yes a PR would be much appreciated! And yes, I believe this is harmless.

**Comment by mlaily** [comment]
I know this issue is old, but I experienced it anyway today so I was glad to find it and I hope what I found can help others...

I found it weird to have to install an extension in order to resolve an error, so I dug up a little bit more.

**The root cause** of the error seems to be the `Fable.Elmish.Debugger` nuget (https://elmish.github.io/debugger/), and its javascript dependency: the `remotedev` npm package.

After removing both from my Fable app, the error disappeared.

I'm not exactly sure what I'm losing beside a few yarn warnings but I'll probably won't miss it too much!
